x/ibc-rate-limit/ibc_middleware_test.go

```diff
@@ -0,0 +1,275 @@
+	suite.chainA.MoveEpochsToTheFuture()
+	suite.chainB = &osmosisibctesting.TestChain{
+		TestChain: suite.coordinator.GetChain(ibctesting.GetChainID(2)),
+	}
+	suite.path = NewTransferPath(suite.chainA, suite.chainB)
+	suite.coordinator.Setup(suite.path)
+}
+
+// Helpers
+func (suite *MiddlewareTestSuite) NewValidMessage(forward bool, amount sdk.Int) sdk.Msg {
+	var coins sdk.Coin
+	var port, channel, accountFrom, accountTo string
+
+	if forward {
+		coins = sdk.NewCoin(sdk.DefaultBondDenom, amount)
+		port = suite.path.EndpointA.ChannelConfig.PortID
+		channel = suite.path.EndpointA.ChannelID
+		accountFrom = suite.chainA.SenderAccount.GetAddress().String()
+		accountTo = suite.chainB.SenderAccount.GetAddress().String()
+	} else {
+		coins = transfertypes.GetTransferCoin(
+			suite.path.EndpointB.ChannelConfig.PortID,
+			suite.path.EndpointB.ChannelID,
+			sdk.DefaultBondDenom,
+			sdk.NewInt(1),
+		)
+		coins = sdk.NewCoin(sdk.DefaultBondDenom, amount)
+		port = suite.path.EndpointB.ChannelConfig.PortID
+		channel = suite.path.EndpointB.ChannelID
+		accountFrom = suite.chainB.SenderAccount.GetAddress().String()
+		accountTo = suite.chainA.SenderAccount.GetAddress().String()
+	}
+
+	timeoutHeight := clienttypes.NewHeight(0, 100)
+	return transfertypes.NewMsgTransfer(
+		port,
+		channel,
+		coins,
+		accountFrom,
+		accountTo,
+		timeoutHeight,
+		0,
+	)
+}
+
+func (suite *MiddlewareTestSuite) ExecuteReceive(msg sdk.Msg) (string, error) {
+	res, err := suite.chainB.SendMsgsNoCheck(msg)
+	suite.Require().NoError(err)
+
+	packet, err := ibctesting.ParsePacketFromEvents(res.GetEvents())
+	suite.Require().NoError(err)
+
+	err = suite.path.EndpointA.UpdateClient()
+	suite.Require().NoError(err)
+
+	res, err = suite.path.EndpointA.RecvPacketWithResult(packet)
+	suite.Require().NoError(err)
+
+	ack, err := ibctesting.ParseAckFromEvents(res.GetEvents())
+	return string(ack), err
+}
+
+func (suite *MiddlewareTestSuite) AssertReceiveSuccess(success bool, msg sdk.Msg) (string, error) {
+	ack, err := suite.ExecuteReceive(msg)
+	if success {
+		suite.Require().NoError(err)
+		suite.Require().NotContains(string(ack), "error",
+			"acknoledgment is an error")
+	} else {
+		suite.Require().Contains(string(ack), "error",
+			"acknoledgment is not an error")
+		suite.Require().Contains(string(ack), types.RateLimitExceededMsg,
+			"acknoledgment error is not of the right type")
+	}
+	return ack, err
+}
+
+func (suite *MiddlewareTestSuite) AssertSendSuccess(success bool, msg sdk.Msg) (*sdk.Result, error) {
+	r, err := suite.chainA.SendMsgsNoCheck(msg)
+	if success {
+		suite.Require().NoError(err, "IBC send failed. Expected success. %s", err)
+	} else {
+		suite.Require().Error(err, "IBC send succeeded. Expected failure")
+		suite.ErrorContains(err, types.RateLimitExceededMsg, "Bad error type")
+	}
+	return r, err
+}
+
+func (suite *MiddlewareTestSuite) BuildChannelQuota(name string, duration, send_precentage, recv_percentage uint32) string {
```

Comment:
Asking to learn - where is the value that we apply `send_percentage` and `recv_percentage` to?

Reply:
in the contract. The message adds the quota to the contract which checks `send_percentage` and `recv_percentage` when calculating the capacity for the denom in the channel: https://github.com/osmosis-labs/osmosis/blob/main/x/ibc-rate-limit/contracts/rate-limiter/src/state.rs#L165-L178

Reply:
That makes sense, ty!